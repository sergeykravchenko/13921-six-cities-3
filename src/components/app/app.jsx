import React, {PureComponent} from "react";
import PropTypes from "prop-types";
import {Switch, Route, BrowserRouter} from "react-router-dom";
import {connect} from "react-redux";
import {ActionCreator} from "../../reducer/state/state";
import {getAllOffers} from "../../reducer/data/selectors";
import {getCities, getActiveCity, getOffers} from "../../reducer/state/selectors";
import Main from "../main/main.jsx";
import OfferCard from "../offer-card/offer-card.jsx";

class App extends PureComponent {
  _renderApp() {
    const {
      offers,
      hoveredOffer,
      cities,
      activeCity,
      activeOffer,
      activeSortType,
      handleCityClick,
      handlePlaceTitleClick,
      handleSortTypeClick,
      handleCardHover,
    } = this.props;

    if (activeOffer) {
      return (
        <OfferCard
          offer={activeOffer}
          offers={offers}
          activeCity={activeCity}
          handlePlaceTitleClick={handlePlaceTitleClick}
        />
      );
    }
    return (
      <Main
        offers={offers}
        hoveredOffer={hoveredOffer}
        handlePlaceTitleClick={handlePlaceTitleClick}
        cities={cities}
        activeCity={activeCity}
        activeSortType={activeSortType}
        handleCityClick={handleCityClick}
        handleSortTypeClick={handleSortTypeClick}
        handleCardHover={handleCardHover}
      />
    );
  }

  render() {
    return <BrowserRouter>
      <Switch>
        <Route exact path="/">
          {this._renderApp()}
        </Route>
        <Route exact path="/dev-offer">
          <OfferCard />
        </Route>
      </Switch>
    </BrowserRouter>;
  }
}

App.propTypes = {
  offers: PropTypes.array,
  cities: PropTypes.array,
  hoveredOffer: PropTypes.number,
  activeCity: PropTypes.object,
  activeOffer: PropTypes.object,
  activeSortType: PropTypes.string,
  handleCityClick: PropTypes.func.isRequired,
  handlePlaceTitleClick: PropTypes.func.isRequired,
  handleSortTypeClick: PropTypes.func,
  handleCardHover: PropTypes.func,
};

const mapStateToProps = (state) => {
  return {
    allOffers: getAllOffers(state),
    cities: getCities(state),
    activeCity: getActiveCity(state),
    offers: getOffers(state),
    activeOffer: state.activeOffer,
    hoveredOffer: state.hoveredOffer,
    activeSortType: state.activeSortType,
  };
};

const mapDispatchToProps = (dispatch) => ({
  handleCityClick(activeCity) {
    dispatch(ActionCreator.changeCity(activeCity));
    dispatch(ActionCreator.getOffers(activeCity.name));
  },
  handlePlaceTitleClick(offer) {
    dispatch(ActionCreator.getActiveOffer(offer));
  },
  handleSortTypeClick(type) {
    dispatch(ActionCreator.getActiveSortType(type));
  },
  handleCardHover(id) {
    dispatch(ActionCreator.getHoveredOffer(id));
  }
});

export {App};
export default connect(mapStateToProps, mapDispatchToProps)(App);
