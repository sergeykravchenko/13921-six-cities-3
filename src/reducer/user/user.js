const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

const initialState = {
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  user: null,
};

const ActionType = {
  REQUIRED_AUTHORIZATION: `REQUIRED_AUTHORIZATION`,
  SET_USER: `SET_USER`,
};

const ActionCreator = {
  requireAuthorization: (status) => {
    return {
      type: ActionType.REQUIRED_AUTHORIZATION,
      payload: status,
    };
  },
  setUser: (data) => {
    return {
      type: ActionType.SET_USER,
      payload: data,
    };
  }
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.REQUIRED_AUTHORIZATION:
      return Object.assign({}, state, {
        authorizationStatus: action.payload,
      });

    case ActionType.SET_USER:
      return Object.assign({}, state, {
        user: action.payload,
      });
  }

  return state;
};

const Operation = {
  checkAuth: () => (dispatch, getState, api) => {
    return api.get(`/login`)
      .then(() => {
        dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
      })
      .catch((err) => {
        throw err;
      });
  },

  login: (authData) => (dispatch, getState, api) => {
    return api.post(`/login`, {
      email: authData.email,
      password: authData.password,
    })
      .then((response) => {
        dispatch(ActionCreator.setUser(response.data));
        dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
      });
  },
};

export {
  ActionCreator,
  ActionType,
  AuthorizationStatus,
  Operation,
  reducer,
};
