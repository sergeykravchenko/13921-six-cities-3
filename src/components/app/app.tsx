import * as React from "react";
import {Switch, Route, Router, Redirect} from "react-router-dom";
import {connect} from "react-redux";
import {AuthorizationStatus} from "../../reducer/user/user";
import {getAuthorizationStatus} from "../../reducer/user/selectors";
import PrivateRoute from "../private-route/private-route";
import Main from "../main/main";
import Favorites from "../favorites/favorites";
import OfferCard from "../offer-card/offer-card";
import SignIn from "../sign-in/sign-in";
import history from "../../history";
import {AppRoute} from "../../utils";

const App: React.FC<{isAuthenticated: boolean}> = ({isAuthenticated}) => {
  return <Router
    history={history}
  >
    <Switch>
      <Route exact path={AppRoute.ROOT} component={Main}/>
      <Route exact path={`${AppRoute.OFFER}/:id`} component={OfferCard}/>
      <Route exact path={AppRoute.LOGIN}
        render={() =>
          !isAuthenticated ?
            <SignIn/>
            :
            <Redirect to={AppRoute.ROOT} />
        }
      />
      <PrivateRoute
        exact
        path={AppRoute.FAVORITES}
        isAuthenticated={isAuthenticated}
        render={() => {
          return (<Favorites/>);
        }}
      />
    </Switch>
  </Router>;
};

const mapStateToProps = (state) => {
  return {
    isAuthenticated: getAuthorizationStatus(state) === AuthorizationStatus.AUTH,
  };
};

export {App};
export default connect(mapStateToProps, null)(React.memo(App));
