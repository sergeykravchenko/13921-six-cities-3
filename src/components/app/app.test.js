import React from 'react';
import renderer from 'react-test-renderer';
import {Provider} from "react-redux";
import configureStore from "redux-mock-store";
import {App} from './app.jsx';
import {AuthorizationStatus} from "../../reducer/user/user.js";
import NameSpace from "../../reducer/name-space.js";

const mockStore = configureStore([]);

it(`App renders correctly`, () => {
  const store = mockStore({
    [NameSpace.DATA]: {
      allOffers: [],
    },
    [NameSpace.USER]: {
      authorizationStatus: AuthorizationStatus.NO_AUTH,
      user: null,
    },
    [NameSpace.STATE]: {
      isFetching: true,
    },
  });
  const tree = renderer
    .create(
        <Provider store={store}>
          <App
            isAuthenticated={AuthorizationStatus.NO_AUTH === `NO_AUTH`}
          />
        </Provider>,
        {
          createNodeMock: () => {
            return document.createElement(`div`);
          }})
    .toJSON();
  expect(tree).toMatchSnapshot();
});
