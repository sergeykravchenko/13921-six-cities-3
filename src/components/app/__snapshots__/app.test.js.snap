// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<p>
  Грузим предложения ...
</p>
`;
